File: 2022/day03/ruby/solutionB.rb
#!/bin/env ruby
# frozen_string_literal: true

require 'set'

input = File.read('../input.txt')

# @param {String} input
# @return {Array(String)}
def format_input(input)
  input.split
end

# @param {Array(String)} rucksacks
# @return {Integer}
def solution(rucksacks)
  priorities = []
  (0...rucksacks.length).step(3) do |i|
    first = rucksacks[i].chars.to_set
    second = rucksacks[i + 1].chars.to_set
    third = rucksacks[i + 2].chars.to_set
    priority = first.&(second).&(third).first
    priorities << priority.ord - (priority > 'Z' ? 'a'.ord - 1 : 38)
  end
  priorities.sum
end

print solution(format_input(input))

File: day03/python/solutionB.py
#!/bin/env python

input = open('../input.txt').read()

def format(input: str) -> list[str]:
    return input.split()

def solution(rucksacks:list[str]) -> int:
    priorities = []
    
    for i in range(0, len(rucksacks) - 1, 3):
        first = set(rucksacks[i])
        second = set(rucksacks[i + 1])
        third = set(rucksacks[i + 2])
        item = first.intersection(second).intersection(third).pop()
        priorities.append(ord(item) - (ord('a') - 1 if item > 'Z' else 38))
    
    return sum(priorities)

print(solution(format(input)))
